### PR TITLE
OSD:memory leak in ReplicatedPG.cc

### PR DESCRIPTION
**Pull request by zhengbinh3c**
"return NULL" in "if (p == snapset.clones.end())" can not free the
memory of "*repop"

Signed-off-by: Bin Zheng <zhengbin.08747@h3c.com>

### CODE REVIEW

**Comment by cxwshawn** [comment]
@zhengbinh3c should remove from repop_queue too.

**Comment by zhengbinh3c** [comment]
@cxwshawn I have modified it, thanks

**Comment by zhengbinh3c** [comment]
@liewegas how do you think about this PR?

**Comment by liewegas** [comment]
I looked at this a bit more closely and I'm worried that this will be the only path where we do simple_repop_create() and then cancel it without submitting it.  That'll be a whole separate path we need to make sure continues to work.

Instead, I think we can move the checks to before the create.  The new_snaps calculation can move up, and then a single check can do a std::find if new_snaps.empty() to ensure that the coid.snap is in snapset.clones before continuing.  Then the current check can be converted to an assert.  Does that make sense?

**Comment by zhengbinh3c** [comment]
@liewegas done it,please check

**Comment by zhengbinh3c** [comment]
@liewegas done it! Please check, thanks

**Comment by zhengbinh3c** [comment]
@liewegas is this OK? what else shuld I do?

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -3254,22 +3246,35 @@ ReplicatedPG::RepGather *ReplicatedPG::trim_object(const hobject_t &coid)
       new_snaps.insert(*i);
   }
 
+  vector<snapid_t>::iterator p = snapset.clones.end();
+
+  if (new_snaps.empty())
+  {
+    p = find(snapset.clones.begin(), snapset.clones.end(), coid.snap);
+    if (p == snapset.clones.end())
+    {
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: style nit: please put { at the end of hte previous line, e.g. if (...) {

File: src/osd/ReplicatedPG.cc [hunk 2]
```diff
@@ -3254,22 +3246,35 @@ ReplicatedPG::RepGather *ReplicatedPG::trim_object(const hobject_t &coid)
       new_snaps.insert(*i);
   }
 
+  vector<snapid_t>::iterator p = snapset.clones.end();
+
+  if (new_snaps.empty())
+  {
+    p = find(snapset.clones.begin(), snapset.clones.end(), coid.snap);
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: please use std::find (we have using namespace std currentyl, but that will hopefully get cleaned up someday)

**Comment by liewegas** [comment]
Looks good, just a few style issues.  Thanks!

**Comment by zhengbinh3c** [comment]
@liewegas done!

**Comment by zhengbinh3c** [comment]
@liewegas done!Please handle it, thanks!

**Comment by zhengbinh3c** [comment]
@liewegas done! Please handle it, thanks!

**Comment by zhengbinh3c** [comment]
@liewegas fix it!

**Comment by zhengbinh3c** [comment]
@liewegas is this OK? what else shuld I do?

**Comment by liewegas** [comment]
yep, just need to run it through qa

**Pull request merged**